lib/tempfile.rb

```diff
@@ -150,26 +150,49 @@ def initialize(basename="", tmpdir=nil, mode: 0, **options)
 
     @unlinked = false
     @mode = mode|File::RDWR|File::CREAT|File::EXCL
+    @finalizer_obj = Object.new
+    tmpfile = nil
     ::Dir::Tmpname.create(basename, tmpdir, **options) do |tmpname, n, opts|
       opts[:perm] = 0600
-      @tmpfile = File.open(tmpname, @mode, **opts)
+      tmpfile = File.open(tmpname, @mode, **opts)
       @opts = opts.freeze
     end
-    ObjectSpace.define_finalizer(self, Remover.new(@tmpfile))
+    ObjectSpace.define_finalizer(@finalizer_obj, Remover.new(tmpfile.path))
+    ObjectSpace.define_finalizer(self, Closer.new(tmpfile))
 
-    super(@tmpfile)
+    super(tmpfile)
+  end
+
+  def initialize_dup(other)
+    initialize_copy_iv(other)
+    super(other)
+    ObjectSpace.define_finalizer(self, Closer.new(__getobj__))
+  end
+
+  def initialize_clone(other)
+    initialize_copy_iv(other)
+    super(other)
+    ObjectSpace.define_finalizer(self, Closer.new(__getobj__))
+  end
+
+  private def initialize_copy_iv(other)
+    @unlinked = other.instance_variable_get(:@unlinked)
+    @mode = other.instance_variable_get(:@mode)
+    @opts = other.instance_variable_get(:@opts)
+    @finalizer_obj = other.instance_variable_get(:@finalizer_obj)
```

Comment:
How about `protected` `attr`s?

Reply:
I don't see how visibility of the methods matters, this just copies the instance variables from the original object (receiver of `dup`/`clone`) to the cloned object.  Maybe I missing something?  Can you provide an example?

Reply:
If these instance variables are `protected`, `instance_variable_get` won’t be needed here.

Reply:
Ah, I think you are suggesting that we should add `protected attr_reader` and then call the methods directly?  Is that correct?

Reply:
Yes, it is correct.

Reply:
I made that change.